enso/experiment/rationalized.py

```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
+            for i, token in enumerate(doc):
+                word_vectors[i,:] = token.vector
+            doc_vectors.append(word_vectors)
+        return doc_vectors
+
+    def featurize_x_y(self, X, Y):
+        doc_vectors = []
+        rationale_targets = []
+        targets = []
+        for text, (rationales, class_target) in zip(X, Y):
+            doc = self.NLP(text, disable=['ner', 'tagger', 'textcat'])
+            word_vectors = np.zeros((len(doc), 300))
+            rationale_one_hot = np.zeros((len(doc), self.n_classes))
+            target_one_hot = self.target_encoder.transform([class_target])[0]
+            for i, token in enumerate(doc):
+                word_vectors[i,:] = token.vector
+                if self._token_in_rationales(token, rationales):
+                    rationale_one_hot[i,:] = target_one_hot
+
+            doc_vectors.append(word_vectors)
+            rationale_targets.append(rationale_one_hot)
+            targets.append(target_one_hot)
+
+        # (batch, seq, hidden) * (batch, seq, one_hot)
+        stacked_docs = np.vstack(doc_vectors)
+        stacked_targets = np.vstack(rationale_targets)
+
+        rationale_proto = np.einsum('sh,st->ht', stacked_docs, stacked_targets)
+        rationale_proto /= np.einsum('st->t', stacked_targets)
+
+        return doc_vectors, rationale_targets, targets, rationale_proto
+        
+      
+    def fit(self, X, Y):
+        # No rationale means everything is a rationale
+        for x, y in zip(X, Y):
+            if not y[0]:
+                y[0].append({"start": 0, "end": len(x), "label": y[1], "text": x})
+       
+        self.target_encoder.fit([y[1] for y in Y])
+        self.n_classes = len(self.target_encoder.classes_)
+
+        doc_vectors, rationale_targets, targets, proto = self.featurize_x_y(X, Y)
+        train = self.train_iter(doc_vectors, rationale_targets, targets, batch_size=2, n_epochs=500)
+
+        model_fn = lambda x, length_mask: TinyNet(kernel_size=1, n_classes=self.n_classes, rationale_proto=proto)(x, length_mask)
+        model = hk.transform(model_fn)
+        rng = jax.random.PRNGKey(0)
+        sample_data_point = next(train)
+        params = model.init(rng, sample_data_point[0], sample_data_point[-1])
+
+        opt_init, opt_update, get_params = optimizers.adam(1e-3)
+        opt_state = opt_init(params)
+        
+        @jax.jit
+        def loss(
+            params: hk.Params,
+            inputs: np.ndarray,
+            rationale_targets: np.ndarray,
+            targets: np.ndarray,
+            length_mask: np.ndarray,
+        ):
+            batch_size = inputs.shape[0]
+
+            rationale_log_probs, clf_log_probs = model.apply(params, x=inputs, length_mask=length_mask)
+            rationale_xent = -jnp.mean(rationale_targets * rationale_log_probs * jnp.expand_dims(length_mask, -1))
+            clf_xent = -jnp.mean(targets * clf_log_probs)
+            return rationale_xent + clf_xent
+
+        @jax.jit
```

Comment:
There is definitely a better way to do this that doesn't involve inlined functions but jax has some awk requirements that make this hard.